### PR TITLE
Update to rails 6

### PR DESCRIPTION
**Pull request by callumknights**
Update publishing-api to rails 6.

See individual commit messages for more detail.

Trello: https://trello.com/c/6hnEg6wT/1820-5-upgrade-publishing-api-to-rails-6

Major note is that this work is somewhat unfinished, some restructuring needs to be done 
for Zeitwerk to work correctly. This will be written into a separate card.

### CODE REVIEW

File: app/queries/live_edition_blocking_draft_edition.rb [hunk 1]
```diff
@@ -12,10 +12,9 @@ def self.call(content_id, base_path, document_type)
 
       conflicts = Edition.with_document
         .where(base_path: base_path, content_store: :live)
-        .where.not(
-          "documents.content_id": content_id,
-          document_type: SubstitutionHelper::SUBSTITUTABLE_DOCUMENT_TYPES,
-        ).pluck(:id)
+        .where.not("documents.content_id": content_id)
```
Comment (thomasleese) [inline app/queries/live_edition_blocking_draft_edition.rb:15]:
```suggestion
        .where.not(documents: { content_id: content_id })
```

I think this is equivalent, which might be nicer to read.

File: config/application.rb [hunk 2]
```diff
@@ -18,7 +22,8 @@
 module PublishingAPI
   class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
-    config.load_defaults 5.1
+    config.load_defaults 6.0
+    config.autoloader = :classic
```
Comment (thomasleese) [inline config/application.rb:26]: It might be worth adding a `FIXME` comment here explaining why this was done, and how to fix it.

File: Gemfile.lock [hunk 3]
```diff
@@ -0,0 +1,1431 @@
...
+  spring-commands-rspec
+  stackprof
+  timecop
+  web-console (~> 3)
+  webmock
+  whenever (= 1.0.0)
+  with_advisory_lock (~> 4.6)
+
+BUNDLED WITH
+   1.17.2
```
Comment (cbaines) [inline Gemfile.lock:1431]: How come this commit is adding a `Gemfile.lock` file?

File: config/application.rb [hunk 4]
```diff
@@ -31,8 +32,10 @@ class Application < Rails::Application
     # -- all .rb files in that directory are automatically loaded.
 
     config.eager_load_paths << "#{config.root}/app"
+    config.eager_load_paths += Dir["#{config.root}/app/queries"]
+    config.eager_load_paths += Dir["#{config.root}/app/commands"]
+    config.eager_load_paths += Dir["#{config.root}/app/presenters"]
```
Comment (cbaines) [inline config/application.rb:38]: Are these changes required to use classic mode?

File: app/queries/live_edition_blocking_draft_edition.rb [hunk 5]
```diff
@@ -12,7 +12,7 @@ def self.call(content_id, base_path, document_type)
 
       conflicts = Edition.with_document
         .where(base_path: base_path, content_store: :live)
-        .where.not("documents.content_id": content_id)
+        .where.not(documents: { content_id: content_id })
```
Comment (cbaines) [inline app/queries/live_edition_blocking_draft_edition.rb:15]: This change doesn't seem to belong in this commit? At least it doesn't have anything to do with `index_with`.

File: config/application.rb [hunk 6]
```diff
@@ -24,6 +24,7 @@ class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 6.0
     config.autoloader = :classic
+    # FIXME Autoloader is only set to classic while restructuring work is undertaken.
```
Comment (cbaines) [inline config/application.rb:27]: This comment should be moved to the relevant commit.

File: Gemfile.lock (hunk 3 above)
Comment (callumknights) [reply Gemfile.lock:1431]: I had issues with my branch build because of a dependabot bump on master, simplest solve was just to remove it and re-bundle

**Comment by cbaines** [comment]
What's in progress with the "WIP Review new framework defaults" commit?

File: config/application.rb (hunk 4 above)
Comment (callumknights) [reply config/application.rb:38]: No they aren't so could comment them out I guess. Make it easier to add them back in when the work has been done.

File: app/queries/live_edition_blocking_draft_edition.rb (hunk 5 above)
Comment (callumknights) [reply app/queries/live_edition_blocking_draft_edition.rb:15]: I guess you're correct, I should move that to a relevant commit.

File: config/application.rb (hunk 6 above)
Comment (callumknights) [reply config/application.rb:27]: Same as above, I'll move it's own commit.

File: Gemfile.lock (hunk 3 above)
Comment (thomasleese) [reply Gemfile.lock:1431]: I think if you rebase against master and on f1db7a271f06475a20a85aca1b1696fc9afc695e, you can run `bundle install` which will generate the latest `Gemfile.lock`, and then move on to the next commit and for each one where you get conflicts you can run `bundle install` to keep the `Gemfile.lock` file up to date. If there are conflicts, I think `bundle install` will complain and make you first delete `Gemfile.lock` before running `bundle install`.

File: config/environments/development.rb [hunk 7]
```diff
@@ -59,5 +59,7 @@
       environment variable.
     MESSAGE
   end
+
+  config.hosts << "publishing-api.dev.gov.uk"
```
Comment (ChrisBAshton) [inline config/environments/development.rb:61]: The link "Issue on govuk-docker" in this commit links to a PR in the 'support' repo. Could you fix it?

File: Gemfile [hunk 8]
```diff
@@ -1,7 +1,6 @@
 source "https://rubygems.org"
 
 # Bundle edge Rails instead: gem "rails", github: "rails/rails"
-gem "arel", "~> 9"
```
Comment (ChrisBAshton) [inline Gemfile:4]: Minor: would be nice to provide a link in the commit message, showing where it says Arel is included in ActiveRecord.

File: Gemfile [hunk 9]
```diff
@@ -36,6 +36,7 @@ gem "whenever", "1.0.0", require: false
 gem "with_advisory_lock", "~> 4.6"
 
 group :development do
+  gem "listen"
```
Comment (ChrisBAshton) [inline Gemfile:39]: Can you explain in the commit message why you've added this gem?

File: app/queries/live_edition_blocking_draft_edition.rb [hunk 10]
```diff
@@ -12,10 +12,9 @@ def self.call(content_id, base_path, document_type)
 
       conflicts = Edition.with_document
         .where(base_path: base_path, content_store: :live)
-        .where.not(
-          "documents.content_id": content_id,
-          document_type: SubstitutionHelper::SUBSTITUTABLE_DOCUMENT_TYPES,
-        ).pluck(:id)
+        .where.not("documents.content_id": content_id)
+        .where.not(document_type: SubstitutionHelper::SUBSTITUTABLE_DOCUMENT_TYPES)
+        .pluck(:id)
```
Comment (ChrisBAshton) [inline app/queries/live_edition_blocking_draft_edition.rb:17]: Nice 👍

File: config/environments/development.rb [hunk 11]
```diff
@@ -10,7 +10,7 @@
   config.cache_classes = false
 
   # Do not eager load code on boot.
-  config.eager_load = false
+  config.eager_load = true
```
Comment (ChrisBAshton) [inline config/environments/development.rb:13]: This doesn't look like a framework default, maybe move this change to a separate commit where you explain why it's done?

File: Gemfile.lock (hunk 3 above)
Comment (cbaines) [reply Gemfile.lock:1431]: The "Remove Arel gem" commit now does look better, but there are too many changes in the Gemfile.lock, only arel should be changing. Can you take a look?

File: config/environments/development.rb (hunk 7 above)
Comment (callumknights) [reply config/environments/development.rb:61]: To me this reads more like a prior art example, so maybe the description of that link could do with changing but I think it's still the right thing to link to. In fact following through from that commit I think you were the first person to encounter this back in august with link-checker-api?

File: Gemfile (hunk 8 above)
Comment (callumknights) [reply Gemfile:4]: Good catch, I'll dig out something from the docs.

File: config/environments/development.rb (hunk 7 above)
Comment (ChrisBAshton) [reply config/environments/development.rb:61]: Yes I think I was! Updating the link description would be fine.

File: Gemfile.lock (hunk 3 above)
Comment (callumknights) [reply Gemfile.lock]: This should look much better now

Comment (cbaines) [reply Gemfile.lock]: Indeed it does, good stuff :+1:

File: config/environments/development.rb (hunk 11 above)
Comment (callumknights) [reply config/environments/development.rb:13]: Having had a chat with Deb this is part of Zeitwerk's autoloading based on naming convention https://github.com/alphagov/signon/pull/1356/commits/46921427a34dca36e2fb4dc334b324ee4ac73424
With that in mind I've temporarily disabled this as zeitwerk is a lot of work to implement in publishing-api and will be brought in in a separate ticket, I'll make sure to add this as a 'first thing to do' in the ticket.

File: Gemfile (hunk 9 above)
Comment (callumknights) [reply Gemfile:39]: This should be explained now 😄

Comment (ChrisBAshton) [reply Gemfile:39]: Great!
